guppylang/checker/expr_checker.py

```diff
@@ -231,14 +234,44 @@ def visit_Call(self, node: ast.Call, ty: Type) -> tuple[ast.expr, Subst]:
             if isinstance(defn, CallableDef):
                 return defn.check_call(node.args, ty, node, self.ctx)
 
-        # Otherwise, it must be a function as a higher-order value
+        # Otherwise, it must be a function as a higher-order value - something
+        # whose type is either a FunctionType or a Tuple of FunctionTypes
         if isinstance(func_ty, FunctionType):
             args, return_ty, inst = check_call(func_ty, node.args, ty, node, self.ctx)
             check_inst(func_ty, inst, node)
             node.func = instantiate_poly(node.func, func_ty, inst)
             return with_loc(node, LocalCall(func=node.func, args=args)), return_ty
-        elif f := self.ctx.globals.get_instance_func(func_ty, "__call__"):
-            return f.check_call(node.args, ty, node, self.ctx)
+
+        if isinstance(func_ty, TupleType) and (
+            function_elements := parse_function_tensor(func_ty)
+        ):
+            assert isinstance(function_elements, list)
+            if any(f.parametrized for f in function_elements):
+                raise GuppyTypeError(
+                    "Polymorphic functions in tuples are not supported", node.func
+                )
+
+            tensor_ty = function_tensor_signature(function_elements)
+
+            processed_args, subst, inst = check_call(
+                tensor_ty, node.args, tensor_ty.output, node.func, self.ctx
+            )
```

Comment:
I think if you do this instead you can get rid of the unification below
```suggestion
            processed_args, subst, inst = check_call(
                tensor_ty, node.args, ty, node, self.ctx
            )
            assert len(inst) == 0
            return return with_loc(node, TensorCall(...)), subst
```

Reply:
Also notice the updated location `node.func` to `node`. I think this should improve the error location in some of the golden tests